src/Interpreters/HashJoin.cpp

```diff
@@ -718,14 +768,14 @@ bool HashJoin::addJoinedBlock(const Block & source_block, bool check_limits)
                     save_nullmap |= (*null_map)[i];
             }
 
-            auto join_mask_col = JoinCommon::getColumnAsMask(block, onexprs[onexpr_idx].condColumnNames().second);
+            auto join_mask_col = JoinCommon::getColumnAsMask(source_block, onexprs[onexpr_idx].condColumnNames().second);
             /// Save blocks that do not hold conditions in ON section
             ColumnUInt8::MutablePtr not_joined_map = nullptr;
-            if (!multiple_disjuncts && isRightOrFull(kind) && !join_mask_col.isConstant())
+            if (!multiple_disjuncts && isRightOrFull(kind) && join_mask_col.hasData())
```

Comment:
It's for conditions like `... JOIN ON t1.a = t2.a AND t2.b > 0`. It was handled incorrectly when we have constant. Get rid of handling constant in JoinMask, this optimization doesn't worth to complicate the code.

Reply:
Please add a test for this case

Reply:
I checked, actually we don't support constants with other conditions in JOIN ON (`JOIN ON t1.a = t2.a AND 1`) on syntax level, that's why bug was not visible. I've added test cases expecting an error.